go-controller/pkg/crd/clusteruserdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	udnv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterUserDefinedNetwork describe network request for a shared network across namespaces.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=clusteruserdefinednetworks,scope=Cluster
+// +kubebuilder:singular=clusteruserdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type ClusterUserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ClusterUserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status ClusterUserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+	// NamespaceSelector Label selector for which namespace network should be available for.
+	// +kubebuilder:validation:Required
+	// +required
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector"`
+
+	// Template is direct specification of UserDefinedNetwork.
+	// +kubebuilder:validation:Required
+	// +required
+	Template UserDefinedNetworkTemplateSpec `json:"template"`
```

Comment:
Having the "template" layer enable extension for other fields then spec, for example: annotation/labels etc..
I followed other k8s object example such as the Demployment,DaemonSet, they have a template and the actual spec inside it.

WDYT?

Reply:
Given @ormergi 's answer, I'd go with template.

Reply:
so the difference is that usually a template contains the exact spec of the object that will be created, e.g. daemonset and deployment template is a pod spec. When you see cluster UDN template, it gives a feeling that namespaced UDNs will be created from that template, which is not the case. So unless we can put a spec of an existing CRD to the "template" section, I don't think it follows best practices.
Also, namespace and cluster scoped UDNs are not that different in this sense, as both create NADs, so if more NAD-realted functionality is needed, it will likely be done for both namespaced and cluster scoped UDNs.

Reply:
Could this `template` attribute be a leftover from when we wanted this cluster wide UDN to template individual namespace UDNs ? 

If the motivation for having this attribute is **only** to pass annotations / labels, I think it's not good enough (motivation). We can simply carry-over those from the cluster wide UDN to the NAD, and not have to expose that in the API.

Reply:
after a team meeting discussion, we agreed that template is not required if we don't want to create namespace copies of UDN (which I think we agreed not to do).

Reply:
Done